lib/IRGen/GenConstant.cpp

```diff
@@ -0,0 +1,140 @@
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements IR generation for constant values.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/Constants.h"
+
+#include "GenConstant.h"
+#include "GenStruct.h"
+#include "GenTuple.h"
+#include "TypeInfo.h"
+#include "swift/Basic/Range.h"
+
+using namespace swift;
+using namespace irgen;
+
+llvm::Constant *irgen::emitConstantInt(IRGenModule &IGM,
+                                       IntegerLiteralInst *ILI) {
+  APInt value = ILI->getValue();
+  BuiltinIntegerWidth width
+    = ILI->getType().castTo<BuiltinIntegerType>()->getWidth();
+
+  // The value may need truncation if its type had an abstract size.
+  if (!width.isFixedWidth()) {
+    assert(width.isPointerWidth() && "impossible width value");
+
+    unsigned pointerWidth = IGM.getPointerSize().getValueInBits();
+    assert(pointerWidth <= value.getBitWidth()
+           && "lost precision at AST/SIL level?!");
+    if (pointerWidth < value.getBitWidth())
+      value = value.trunc(pointerWidth);
+  }
+
+  return llvm::ConstantInt::get(IGM.LLVMContext, value);
+}
+
+llvm::Constant *irgen::emitConstantFP(IRGenModule &IGM, FloatLiteralInst *FLI) {
+  return llvm::ConstantFP::get(IGM.LLVMContext, FLI->getValue());
+}
+
+llvm::Constant *irgen::emitAddrOfConstantString(IRGenModule &IGM,
+                                                StringLiteralInst *SLI) {
+  switch (SLI->getEncoding()) {
+  case StringLiteralInst::Encoding::UTF8:
+    return IGM.getAddrOfGlobalString(SLI->getValue());
+
+  case StringLiteralInst::Encoding::UTF16: {
+    // This is always a GEP of a GlobalVariable with a nul terminator.
+    auto addr = IGM.getAddrOfGlobalUTF16String(SLI->getValue());
+
+    // Cast to Builtin.RawPointer.
+    return llvm::ConstantExpr::getBitCast(addr, IGM.Int8PtrTy);
+  }
+
+  case StringLiteralInst::Encoding::ObjCSelector:
+    llvm_unreachable("cannot get the address of an Objective-C selector");
+  }
+  llvm_unreachable("bad string encoding");
+}
+
+static llvm::Constant *emitConstantValue(IRGenModule &IGM, SILValue Operand) {
+  if (auto *Elem = dyn_cast<StructInst>(Operand))
+    return emitConstantStruct(IGM, Elem);
+  else if (auto *Elem = dyn_cast<TupleInst>(Operand))
+    return emitConstantTuple(IGM, Elem);
+  else if (auto *ILI = dyn_cast<IntegerLiteralInst>(Operand))
+    return emitConstantInt(IGM, ILI);
+  else if (auto *FLI = dyn_cast<FloatLiteralInst>(Operand))
+    return emitConstantFP(IGM, FLI);
+  else if (auto *SLI = dyn_cast<StringLiteralInst>(Operand))
+    return emitAddrOfConstantString(IGM, SLI);
+  else
+    llvm_unreachable("Unsupported SILInstruction in static initializer!");
+}
+
+template <typename InstTy, typename NextIndexFunc>
+llvm::Constant *emitConstantStructOrTuple(IRGenModule &IGM, InstTy Inst,
+                                          NextIndexFunc NextIndex) {
```

Comment:
I'm not thrilled that this really is a NextIndexFunc, i.e. it has to be called in order, but given that it is, ok.

IRGen convention is to lowercase local variable names; please maintain that.